src/frontend/src/optimizer/plan_rewriter/iceberg_source_rewriter.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::catalog::ICEBERG_SEQUENCE_NUM_COLUMN_NAME;
+use risingwave_common::util::iter_util::ZipEqFast;
+use risingwave_connector::source::iceberg::IcebergSplitEnumerator;
+use risingwave_connector::source::{ConnectorProperties, SourceEnumeratorContext};
+use risingwave_pb::batch_plan::iceberg_scan_node::IcebergScanType;
+
+use crate::error::Result;
+use crate::expr::{ExprImpl, ExprType, FunctionCall, InputRef};
+use crate::optimizer::plan_node::generic::GenericPlanNode;
+use crate::optimizer::plan_node::{LogicalIcebergScan, LogicalJoin, LogicalSource};
+use crate::optimizer::PlanRef;
+use crate::utils::{Condition, FRONTEND_RUNTIME};
+
+#[derive(Debug, Clone, Default)]
+pub struct IcebergMergeOnReadRewriter {}
```

Comment:
Currently, our optimizer rule supports returning an error after https://github.com/risingwavelabs/risingwave/pull/19381 I think we can do a refactor later